src/routes/assistants/+page.server.ts

```diff
@@ -34,6 +35,7 @@ export const load = async ({ url, locals }) => {
 		...(modelId && { modelId }),
 		...(!createdByCurrentUser && { userCount: { $gt: 1 } }),
 		...(user ? { createdById: user._id } : { featured: true }),
+		...(query && { name: { $regex: query, $options: "i" } }),
```

Comment:
For this filter change, I've created this index https://github.com/huggingface/chat-ui/blob/assistants_search/src/lib/server/database.ts#L88-L89

Please let me know if it is enough (cc: @coyotte508 )

Reply:
If you want to efficiently search, besides the `$text` index (more suited to full text), if you want to do it in mongo you need prefixed case-sensitive regexes, and store search tokens.

Eg we have this in moon-landing:

```
generateSearchTokens(str: string): string[] {
		const lcase = str.toLowerCase().trim();

		return [
			...new Set(
				[
					lcase.replace(/[-._\s/]/g, ""),
					...lcase.split("/").map(x => x.replace(/[-._\s]/g, "")),
					...lcase.replace(/[-.\s_/]/g, " ").split(" "),
					...lcase
						.replace(/[-\s_/]/g, " ")
						.split(" ")
						.map(x => x.replace(/\./g, "")),
					/// ^ do not split on `.` but still remove them from search tokens (simpler for search)
					/// allows for searches like "2.7B" to return things like "EleutherAI/gpt-neo-2.7B"
				].filter(x => !!x)
			),
		];
	}
```

You can apply this to the name - probably tweaking it a bit since the strings we process are a bit different - and store the search tokens.

Then create an index on `searchTokens: 1, userCount: -1`

Then the search, you do `{ searchTokens: { $all: q.split(/[-._/\s]+/).map(makeRegex) } }` with

```
				const makeRegex = (s: string) => new RegExp("^" + Utils.escapeForRegExp(s));
```

(which prefixes the regex by using `^`)

At least that's how we do it for the hub. Alterative to this and `$text` is using meilisearch :sweat_smile:

Reply:
Right now, we plan to only filter on `assistant.name`. Would this filter `(query && { name: { $regex: query, $options: "i" } }),` too slow considering that we have 10k+ assistants atm ?

We can definitely create/use moon-landing-like `searchTokens` if we also want to include `assistant.description` besides `assistant.name`

Reply:
`searchTokens` are suitable for assistant.name. If we included description we'd have to move to a `$text` index instead.

We use `searchTokens` on the hub to search org names, model names, user names etc

> Would this filter (query && { name: { $regex: query, $options: "i" } }), be too slow considering that we have 10k+ assistants atm ?

As long as the follow up PR is not forgotten - but may as well do it right the first time

Reply:
@gary149 wdyt ?

Reply:
@coyotte508 the question is:
1. Can we start with this simple `name: { $regex: query, $options: "i" }` only on `assistant.name` without having to create `searchTokens`? By can we, I mean: would the performance be good enough for the number of assistants we have? Or we don't know how the experience is gonna be without trying it empirically?
2. However, if we have to use `searchTokens`, then it would make sense to add `assistant.description` to `searchTokens` as well as `assistant.name`

Reply:
IMO:

- Search name only: use `searchTokens`
- Search name & description: use a [text index](https://www.mongodb.com/docs/manual/core/indexes/index-types/index-text/create-text-index/#std-label-compound-text-index-example)

Reply:
Ok let's use `searchTokens` then.

Reply:
ready for review